### PR TITLE
Do not import the VSWorkspace during Roslyn package initialization

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs [hunk 1]
```diff
@@ -22,7 +21,12 @@ namespace Microsoft.CodeAnalysis.Remote
     internal sealed class SolutionChecksumUpdater
     {
         private readonly Workspace _workspace;
-        private readonly IGlobalOperationNotificationService _globalOperationService;
+
+        /// <summary>
+        /// We're not at a layer where we are guaranteed to have an IGlobalOperationNotificationService.  So allow for
+        /// it being null.
+        /// </summary>
+        private readonly IGlobalOperationNotificationService? _globalOperationService;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Remote/SolutionChecksumUpdater.cs:29]: IGlobalOperationNotificationService has been made into a normal mef service (no longer a workspace service).  It is not exported from the core layers.  Instead, there is an export at teh VS, Remote, and Test layers.

File: src/VisualStudio/Core/Def/RoslynPackage.cs [hunk 2]
```diff
@@ -152,16 +151,17 @@ protected override async Task InitializeAsync(CancellationToken cancellationToke
             // Ensure the options persisters are loaded since we have to fetch options from the shell
             LoadOptionPersistersAsync(this.ComponentModel, cancellationToken).Forget();
 
-            _workspace = this.ComponentModel.GetService<VisualStudioWorkspace>();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/RoslynPackage.cs:155]: no more creating hte workspace (which creates its entire mef composition on this thread).

File: src/VisualStudio/Core/Def/RoslynPackage.cs [hunk 3]
```diff
@@ -323,7 +323,8 @@ private void ReportSessionWideTelemetry()
 
         private void DisposeVisualStudioServices()
         {
-            _workspace?.Services.GetRequiredService<VisualStudioMetadataReferenceManager>().DisconnectFromVisualStudioNativeServices();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/RoslynPackage.cs:326]: VisualStudioMetadataReferenceManager is already a normal MEF service.  so no need for a workspace to get it.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski this is ready for review.

File: src/VisualStudio/Core/Def/RoslynPackage.cs [hunk 4]
```diff
@@ -152,16 +151,17 @@ protected override async Task InitializeAsync(CancellationToken cancellationToke
...
-
             await InitializeColorsAsync(cancellationToken).ConfigureAwait(true);
 
             // load some services that have to be loaded in UI thread
             LoadComponentsInUIContextOnceSolutionFullyLoadedAsync(cancellationToken).Forget();
 
-            _solutionEventMonitor = new SolutionEventMonitor(_workspace);
+            // We are at the VS layer, so we know we must be able to get the IGlobalOperationNotificationService here.
+            var globalNotificationService = this.ComponentModel.GetService<IGlobalOperationNotificationService>();
+            Contract.ThrowIfNull(globalNotificationService);
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/RoslynPackage.cs:161]: There's an Assumes.Present() call you can also use out of the VS asserts library that includes the typename, which is otfen used at the VS layer for service things like this.

File: src/VisualStudio/Core/Def/RoslynPackage.cs (hunk 3 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/RoslynPackage.cs:326]: Nice!

File: src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs [hunk 5]
```diff
@@ -105,7 +105,8 @@ internal sealed class ExternalErrorDiagnosticUpdateSource : IDiagnosticUpdateSou
             _diagnosticService = diagnosticService;
             _buildOnlyDiagnosticsService = _workspace.Services.GetRequiredService<IBuildOnlyDiagnosticsService>();
 
-            _notificationService = _workspace.Services.GetRequiredService<IGlobalOperationNotificationService>();
+            // We are in the VS layer, so getting the IGlobalOperationNotificationService must succeed.
+            _notificationService = _workspace.Services.SolutionServices.ExportProvider.GetExports<IGlobalOperationNotificationService>().Single().Value;
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs:109]: It looks like the constructor is run here:

https://github.com/dotnet/roslyn/blob/f26bc3080b6e26c196ade50eb537c6b72e769c35/src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl.cs#L166

So you could just pass this through the constructor which then avoids having to grab the ExportProvider from SolutionServices, which seems a bit sketchy.

**Review by jasonmalinowski — APPROVED** [review]
I admit I'm surprised SolutionServices has an ExportProvider that's accessible in some of those places, but most of this is going away soon with solution crawler anyways so I won't complain!

File: src/VisualStudio/Core/Def/RoslynPackage.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/RoslynPackage.cs]: Done.

File: src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/TaskList/ExternalErrorDiagnosticUpdateSource.cs:109]: will do!

**Review by genlu — APPROVED** [review]

**Pull request merged**